data/json/effects_on_condition/dream_eocs.json

```diff
@@ -5,6 +5,103 @@
+    "//": "Prints a flavour text message to the log about being afraid of sleep if the character has entered the tired status, has the trubbled sleeper trait, and has not taken preventative medication. The EOC’s set to recur every 8 to 20 minutes.",
+    "global": true,
+    "recurrence": [ "8 minutes", "20 minutes" ],
+    "condition": {
+      "and": [
+        { "math": [ "u_val('fatigue')", ">", "191" ] },
+        { "u_has_trait": "bad_dreams" },
+        { "not": { "u_has_effect": "took_prozac" } }
+      ]
+    },
+    "effect": [ { "u_message": "troubled_sleep_fatigue_messages", "snippet": true, "type": "bad" } ]
+  },
+  {
+    "type": "effect_on_condition",
+    "id": "EOC_TROUBLED_SLEEP_MESSAGE",
+    "//": "prints a message to the log upon the character falling asleep with the troubled sleeper trait. The text changes based on whether the player has taken preventative medication or not. Primarily intended for immersion and to let the player know if their medication, should they have taken it, is going to work to repress nightmares.",
+    "eoc_type": "EVENT",
+    "required_event": "character_falls_asleep",
+    "condition": { "u_has_trait": "bad_dreams" },
+    "effect": [
+      {
+        "if": { "u_has_effect": "took_prozac" },
+        "then": {
+          "u_message": "Surrendering yourself to the arms of sleep, you hope your medication will let you have a good night’s rest.",
+          "type": "good"
+        },
+        "else": {
+          "u_message": "As your eyes close, unease brews in your mind, anxiety swirling in your gut.  You know the nightmares won't leave you at peace.",
+          "type": "bad"
+        }
+      }
+    ]
+  }
+  {
+    "type": "effect_on_condition",
+    "id": "EOC_TROUBLED_SLEEP_SOLDIER_NIGHTMARES",
+    "//": "Generates a military-specific nightmare for players with the troubled sleeper trait and a soldier profession upon the character waking up, presents a pop-up with the nightmare, and provides a mood debuff. The EOC’s set to not fire if the player has taken preventative medication.",
+    "eoc_type": "EVENT",
+    "required_event": "character_wakes_up",
+    "condition": {
+      "and": [
+        { "math": [ "dream_counter", "==", "0" ] },
+        { "u_has_trait": "bad_dreams" },
+        { "not": { "u_has_effect": "took_prozac" } },
+        {
+          "or": {
+            "u_profession": "recruit",
+            "u_profession": "rifleman",
+            "u_profession": "specops",
+            "u_profession": "mil_marksman",
+            "u_profession": "mil_auto_rifleman",
+            "u_profession": "mil_grenadier",
+            "u_profession": "mil_breacher",
+            "u_profession": "mil_sniper",
+            "u_profession": "mil_hacker",
+            "u_profession": "mil_undercover",
+            "u_profession": "hazmat_unit",
+            "u_profession": "nco",
+            "u_profession": "combat-mechanic",
+            "u_profession": "combat-engineer"
```

Comment:
I don't know much about EOC / JSON stuff, but I wonder, wouldn't be possible to use some invisible traits as "profession tags" instead? This way mods could make use of this same stuff, for example, some of military-ish MoM professions. That also allow combination of nightmares, like a Military Parametric having both military related and EMT related ones.

OTOH this may cause a "trait-explosion" so ¯\_(ツ)_/¯

Reply:
Yes, it is possible to add invisible traits by setting `"player_display": false`, then you can use EOCs to check for those instead of individual `id`s.

The real downside of doing that isn't the explosion itself (the mutation.json file is close to 10k lines now, however most of those are mutation-related traits iirc), but manually adding each tag to any and all professions that qualify.

Reply:
@GuardianDll, to be honest, chief, you and Renech’s help in getting this PR off its feet on the EOC side of things has been very appreciated. Thanks for the further pointers and optimisations here.
@andrewhr, to be honest, I think I’d rather just make the EOC a player-wide thing rather than making an invizable trait for each and every profession use case. When it comes to mods, I believe they could add their own EOC that basically copies what I’ve made but applies it to their own professions. On the subject, however, of keeping this feature dependent on a trait or not, I’m not sure which way I lean for the time being.

The three options I’m seeing are

- Keep it as a trait, as it is currently.

- Keep it as a trait, but make it come by default in certain scenarios where it would make sense. For example, somebody starting in an overrun military base has probably seen their fair share of shit over the cataclysm, while somebody who’s starting in a safe island cabin might have been relatively free of encountering such events. Down the line, certain events could add the trait to these characters, but we’re now walking into fantasy land ideas.

- Make it so every player gets the fun nightmares and laugh at the complainers when they come huffing about how every sleep makes their character depressed.

Reply:
it is correct, it would be better to keep it as a straight list of profs - bloating profs with unnecessary traits is bad approach, especially when it is used only in one place - here, in this EoC

mods would have no problem to add stuff into the EoC by copying and adding profs they want into the list; if it is a consern for you, please don't think about it; if you have a legit use case, you can move the trait checks to separate EoCs and then use `test_eoc`